pkg/controller/controller.go

```diff
@@ -241,68 +249,44 @@ func (c *Controller) partition(id ID, partitioner WorkPartitioner) {
 				c.mu.Lock()
 				partition, ok = c.partitions[key]
 				if !ok {
-					partition = make(chan ID)
+					partition = make(chan Request)
 					c.partitions[key] = partition
-					go c.processRequests(partition)
+					go c.reconcileRequests(partition, reconciler)
 				}
 				c.mu.Unlock()
 			}
-			partition <- id
+			partition <- Request{
+				ID: id,
+			}
 			return
 		}
 		iteration++
 	}
 }
 
-// processRequests processes requests from the given channel
-func (c *Controller) processRequests(ch chan ID) {
-	c.mu.RLock()
-	reconciler := c.reconciler
-	c.mu.RUnlock()
-
-	for id := range ch {
-		// Reconcile the request. If the reconciliation is not successful, requeue the request to be processed
-		// after the remaining enqueued events.
-		result := c.reconcile(id, reconciler)
-		if result.Requeue.Value != nil {
-			go c.requeueRequest(ch, result.Requeue)
-		}
+// reconcileRequests reconciles requests from the given channel
+func (c *Controller) reconcileRequests(ch chan Request, reconciler Reconciler) {
+	for request := range ch {
+		c.reconcileRequest(request, ch, reconciler)
 	}
 }
 
-// requeueRequest requeues the given request
-func (c *Controller) requeueRequest(ch chan ID, id ID) {
-	ch <- id
-}
-
-// reconcile reconciles the given request ID until complete
-func (c *Controller) reconcile(id ID, reconciler Reconciler) Result {
-
-	iteration := 0
-	var result Result
-	var err error
-	b := backoff.NewExponentialBackOff()
-	b.InitialInterval = initialRetryInterval
-	// MaxInterval caps the RetryInterval
-	b.MaxInterval = maxRetryInterval
-	// Never stops retrying
-	b.MaxElapsedTime = 0
-
-	notify := func(err error, t time.Duration) {
-		log.Infof("An error occurred during reconciliation of %v in iteration %v: %v", id.Value, err, iteration)
-		iteration++
-
+// reconcileRequest reconciles the given request
+func (c *Controller) reconcileRequest(request Request, ch chan Request, reconciler Reconciler) {
+	request.attempt++
+	result, err := reconciler.Reconcile(request.ID)
+	if err != nil {
+		backoffDelay := 10 * time.Millisecond * time.Duration(math.Pow(2, float64(request.attempt)))
```

Comment:
Just curios to see if there is a problem in the back off library?

Reply:
Not particularly. The problem is just that the API isn't really suited to interleaving backoffs for different requests like this. It provides a blocking backoff for one requests, and we need to interleave backoffs for multiple requests without blocking the queue. The `backoff` library provides a blocking backoff, and this is a non-blocking backoff. It could have been used, but it would have required more refactoring of this code.